hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/QueuePath.java

```diff
@@ -116,8 +120,79 @@ public boolean hasParent() {
     return parent != null;
   }
 
+  /**
+   * Creates a new {@code QueuePath} from the current full path as parent, and
+   * the appended child queue path as leaf.
+   * @param childQueue path of leaf queue
+   * @return new queue path made of current full path and appended leaf path
+   */
+  public QueuePath createNewLeaf(String childQueue) {
+    return new QueuePath(getFullPath(), childQueue);
+  }
+
+  /**
+   * Returns an iterator of queue path parts, starting from the highest level
+   * (generally root).
+   * @return queue part iterator
+   */
+  @Override
+  public Iterator<String> iterator() {
+    return Arrays.asList(getFullPath().split(QUEUE_REGEX_DELIMITER)).iterator();
+  }
+
+  /**
+   * Returns an iterator that provides a way to traverse the queue path from
+   * current queue through its parents.
+   * @return queue path iterator
+   */
+  public Iterator<QueuePath> reversePathIterator() {
+    QueuePath initial = this;
+
+    return new Iterator<QueuePath>() {
+      private boolean stop = false;
+      private QueuePath current = initial;
+
+      @Override
+      public boolean hasNext() {
+        return !stop;
+      }
+
+      @Override
+      public QueuePath next() {
+        QueuePath old = current;
+        if (current.hasParent()) {
+          current = new QueuePath(current.getParent());
+        }
+
+        if (!old.hasParent()) {
+          stop = true;
+        }
+
+        return old;
+      }
+    };
+  }
+
   @Override
   public String toString() {
     return getFullPath();
   }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    QueuePath strings = (QueuePath) o;
```

Comment:
Please consider using a bit more expressing name here, like otherPath or something like that, 'strings' is not too informative especially when it is an other QueuePath object and not a string collection :)